### PR TITLE
refactor(ext/fs): boxed deno_fs::FileSystem

### PR DESCRIPTION
**Pull request by dsherret**
1. Boxed `File` and `FileSystem` to allow more easily passing this through the CLI code (as shown within this pr).
2. `StdFileResource` is now `FileResource`. `FileResource` now contains an `Rc<dyn File>`.

Depends on #18959

### CODE REVIEW

File: ext/fs/interface.rs [hunk 1]
```diff
@@ -90,106 +72,38 @@ pub struct FsDirEntry {
...
-
 #[async_trait::async_trait(?Send)]
-pub trait FileSystem: Clone {
-  type File: File;
-
+pub trait FileSystem: Send + Sync {
   fn cwd(&self) -> FsResult<PathBuf>;
   fn tmp_dir(&self) -> FsResult<PathBuf>;
-  fn chdir(&self, path: impl AsRef<Path>) -> FsResult<()>;
+  fn chdir(&self, path: &Path) -> FsResult<()>;
```
Comment (dsherret) [inline ext/fs/interface.rs:79]: Can't be `impl ` if it's boxed.

File: ext/io/fs.rs [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+      gid: unix_or_zero!(gid),
+      rdev: unix_or_zero!(rdev),
+      blksize: unix_or_zero!(blksize),
+      blocks: unix_or_zero!(blocks),
+    }
+  }
+}
+
+#[async_trait::async_trait(?Send)]
+pub trait File {
```
[the remark this answers is not in the document]
Comment (bartlomieju) [reply ext/io/fs.rs:128]: Okay

File: ext/io/fs.rs [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+  // deno_core
+  async fn read_byob(
+    self: Rc<Self>,
+    buf: BufMutView,
+  ) -> FsResult<(usize, BufMutView)>;
+  async fn write_byob(
+    self: Rc<Self>,
+    buf: BufView,
+  ) -> FsResult<deno_core::WriteOutcome>;
+  async fn write_all_byob(self: Rc<Self>, buf: BufView) -> FsResult<()>;
```
[the remark this answers is not in the document]
Comment (bartlomieju) [reply ext/io/fs.rs:188]: Seems fine for the time being

**Review by bartlomieju — APPROVED** [review]
LGTM, I'm not gonna be picky since this code is very much in flux. All tests are passing so it's good for me.

File: ext/io/Cargo.toml
Comment (dsherret) [inline ext/io/Cargo.toml:1]: Will do a PR after that removes ext/io and folds this into ext/fs

File: ext/io/fs.rs [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
```
Comment (lucacasonato) [inline ext/io/fs.rs:188]:
```suggestion
  async fn write(
    self: Rc<Self>,
    buf: BufView,
  ) -> FsResult<deno_core::WriteOutcome>;
  async fn write_all(self: Rc<Self>, buf: BufView) -> FsResult<()>;
```

File: ext/io/fs.rs [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+  async fn utime_async(
+    self: Rc<Self>,
+    atime_secs: i64,
+    atime_nanos: u32,
+    mtime_secs: i64,
+    mtime_nanos: u32,
+  ) -> FsResult<()>;
+
+  // deno_core
+  async fn read_byob(
```
Comment (lucacasonato) [inline ext/io/fs.rs:180]: Add a `async fn read` also. Don't assume the implementation of `read` delegates to `read_byob`.

**Comment by dsherret** [comment]
I'm going to open two PRs after this one:

1. Remove `NodeFs` and just use `deno_fs::FileSystem` everywhere.
2. Ability for `deno_fs::FileSystem` to not be `Send + Sync`

**Comment by dsherret** [comment]
Going to merge this. We can make improvements in follow up prs.

**Pull request merged**